pkgs/by-name/ni/nixos-render-docs/src/nixos_render_docs/redirects.py

```diff
@@ -0,0 +1,172 @@
+    - {"\n    - ".join(self.divergent_redirects)}
+
+    It leads to inconsistent behavior depending on which redirect is applied.
+    Please update doc/redirects.json or nixos/doc/manual/redirects.json!
+""")
+        if self.identifiers_missing_current_outpath:
+            error_messages.append(f"""
+The first element of an identifier's redirects list must denote its current location.
+    The following identifiers violate this rule:
+    - {"\n    - ".join(self.identifiers_missing_current_outpath)}
+
+    If you moved content, add its new location as the first element of the redirects mapping.
+    Please update doc/redirects.json or nixos/doc/manual/redirects.json!
+""") # TODO: automatically detect if you just missed adding a new location, and make a tool to do that for you
+        if self.identifiers_without_redirects:
+            error_messages.append(f"""
+Identifiers present in the source must have a mapping in the redirects file.
+    - {"\n    - ".join(self.identifiers_without_redirects)}
+
+    This can happen when an identifier was added or renamed.
+    Please update doc/redirects.json or nixos/doc/manual/redirects.json!
+""") # TODO: add tooling in the development shell to do that automatically and point to that command
+        if self.orphan_identifiers:
+            error_messages.append(f"""
+Keys of the redirects mapping must correspond to some identifier in the source.
+    - {"\n    - ".join(self.orphan_identifiers)}
+
+    This can happen when an identifier was removed or renamed.
+    Please update doc/redirects.json or nixos/doc/manual/redirects.json!
+""") # TODO: add tooling in the development shell to do that automatically and point to that command
+
+        error_messages.append("NOTE: If your Manual build passes locally and you see this message in CI, you probably need a rebase.")
+        return "\n".join(error_messages)
+
+
+@dataclass
+class Redirects:
+    _raw_redirects: dict[str, list[str]]
+    _redirects_script: str
+
+    _xref_targets: dict[str, XrefTarget] = field(default_factory=dict)
+
+    def validate(self, initial_xref_targets: dict[str, XrefTarget]):
+        """
+        Validate redirection mappings against element locations in the output
+
+        - Ensure semantic correctness of the set of redirects with the following rules:
+          - Identifiers present in the source must have a mapping in the redirects file
+          - Keys of the redirects mapping must correspond to some identifier in the source
+          - All historical content locations must correspond to exactly one identifier
+          - Identifiers must not be identical to any historical location's anchor of the same output path
+          - The first element of an identifier's redirects list must denote its current location.
+        """
+        xref_targets = {}
+        for id, target in initial_xref_targets.items():
+            if id.startswith("opt-") or id.startswith("auto-generated-"):
```

Comment:
```suggestion
            # filter out automatically generated identifiers from module options and library documentation
            if id.startswith("opt-") or id.startswith("auto-generated-"):
```

Reply:
```suggestion
            if id.startswith("opt-") or id.startswith("auto-generated-") or id.startswith("function-library-"):
```

We probably want to catch those, too. Not sure if this removes too much, because I can't grep out of the code if there are such identifiers created manually. Very likely not, since it's [created by nixdoc](https://github.com/nix-community/nixdoc/blob/4fa541a7b270a255caf6bbceeb29d919e6e66704/src/commonmark.rs#L164), but we have to hope that noone wanted to be clever.